versions/development/SPEC.md

```diff
@@ -1218,81 +1270,316 @@ $runtime = 'runtime' $ws* '{' ($ws* $runtime_kv $ws*)* '}'
-    String ubuntu_version
-  }
-  command {
-    python script.py
+task single_image_test {
+  #....
+  runtime {
+    container: "ubuntu:latest"
   }
+```
+
+**Multiple Images**
+```wdl
+task multiple_image_test {
+  #.....
   runtime {
-    docker: "ubuntu:" + ubuntu_version
+    container: ["ubuntu:latest", "gcr.io/standard-images/ubuntu:latest"]
   }
 }
 ```
 
-Most key/value pairs are arbitrary.  However, the following keys have recommended conventions:
 
-#### docker
+#### cpu (**Required**)
 
-Location of a Docker image for which this task ought to be run.  This can have a format like `ubuntu:latest` or `broadinstitute/scala-baseimage` in which case it should be interpreted as an image on DockerHub (i.e. it is valid to use in a `docker pull` command).
+The `cpu` key defines the _minimum_ CPU required for this task, which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of CPU requested (depending on the underlying infrastructure restrictions), however it may ONLY provision more CPU than requested and not less. For example if the wdl requested `cpu: 0.5`, but only discrete values were supported, then the engine might choose to provision `1.0` cpu instead.  Values are expected to be a `Float` (where supported) or an `Int`
 
 ```wdl
-task docker_test {
-  input {
-    String arg
-  }
-  command {
-    python process.py ${arg}
-  }
+task cpu_example {
+  #....	
   runtime {
-    docker: "ubuntu:latest"
+    cpu: 8
   }
 }
+
 ```
 
-#### memory
+#### memory (**Required**)
 
-Memory requirements for this task.  Two kinds of values are supported for this attributes:
+The `memory` key defines the _minimum_ memory required for this task which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of memory requested, however it may ONLY provision more memory then requested and not less. For example, if the wdl requested `1 GB` but only blocks of `4 GB` were available, the engine might choose to provision `4.0 GB` instead.  Two kinds of values are supported for this attribute:
 
 * `Int` - Interpreted as bytes
 * `String` - This should be a decimal value with suffixes like `B`, `KB`, `MB` or binary suffixes `KiB`, `MiB`.  For example: `6.2 GB`, `5MB`, `2GiB`.
 
 ```wdl
 task memory_test {
-  input {
-    String arg
+  #....
+  runtime {
+    memory: "2 GB"
   }
+}
+```
 
-  command {
-    python process.py ${arg}
+#### gpu (**Optional**)
+
+The `gpu` key provides a way to accommodate modern workflows which are increasingly becoming reliant on GPU computations. This key is a `true` or `false` value and simply indicates to the engine that a task requires a GPU to run to completion. A task with this flag set to `true` is guaranteed to only run if a GPU is a available within the runtime environment. It is the responsibility of the engine to check prior to execution whether a GPU is provisionable, and if not, preemptively fail the task.
```

Comment:
Maybe someone with more experience using GPUs can chime in:

* Is "I just need a GPU, I don't care what type" enough information? (ie is it possible to be agnostic about the number, type and driver version required?)
* Is having a GPU a genuine requirement for running such a task
  * ... or is it more a case of "if there isn't a GPU the task will still succeed but will be inefficient?"

Reply:
That is what I am a little naive about, so I will rely on someone else to have more information. I was trying to capture the requirement for a GPU without throwing in all of the provider specific terminology in this part of the block.

Reply:
Just to provide some color here.  

The GPU instance types DNAnexus supports today, on AWS, are the g2 and p3 families. The p3 is the current generation. To allow users to actually make use of these, we install an NVIDIA driver. An alternative is nvidia-docker.

Reply:
@orodeh does the user need to know specifics about this? or is it enough for them to know that a GPU was added, and then allow additional config in the hints section?

Reply:
my motivatin  here was to abstract as much of the specific vendor terminology into the hints section, then have an engine guarantee a gpu was added. the flavor and size of the gpu would need to be either configured in the hints or the engine would need to make a resonable default value here

Reply:
I think it is a good start to say have a GPU flag in the runtime section, and leave the rest for hints.